File: output/ChatGPT/codenet/C++/Python/s003531724.py
import sys
sys.setrecursionlimit(10**6)

def dfs(x, l):
    global cnt, d, ok
    vis[x] = True
    cnt[l] += 1
    d[x] = l
    i = u[x]
    while i:
        if not vis[e[i].to]:
            dfs(e[i].to, (l + e[i].c) % 3)
        elif (l + e[i].c) % 3 != d[e[i].to]:
            ok = False
        i = e[i].pre

n, m = map(int, input().split())
u = [0] * (n + 1)
e = [0] * (m * 2 + 1)
l = 0
for i in range(1, m + 1):
    a, b = map(int, input().split())
    l += 1
    e[l] = [b, u[a], 1]
    u[a] = l
    l += 1
    e[l] = [a, u[b], 2]
    u[b] = l

cnt = [0] * 3
d = [0] * (n + 1)
ans = 0
for i in range(1, n + 1):
    vis = [False] * (n + 1)
    if not vis[i]:
        T = 0
        for j in range(3):
            cnt[j] = 0
        ok = True
        dfs(i, 0)
        s = cnt[0] + cnt[1] + cnt[2]
        if not ok:
            ans += s * s
        else:
            if cnt[0] and cnt[1] and cnt[2]:
                ans += cnt[0] * cnt[1] + cnt[1] * cnt[2] + cnt[2] * cnt[0]
            else:
                ans += T // 2
print(ans)

File: output/ChatGPT/codenet/C++/Java/s003531724.java
import java.util.*;
import java.io.*;

public class s003531724 {
    static final int N = 100010;
    
    static class Edge {
        int to, pre, c;
        public Edge(int to, int pre, int c) {
            this.to = to;
            this.pre = pre;
            this.c = c;
        }
    };
    
    static Edge[] e = new Edge[N << 1];
    static int[] u = new int[N];
    static int l = 0;
    
    static void ins(int a, int b, int c) {
        e[++l] = new Edge(b, u[a], c);
        u[a] = l;
    }
    
    static int n, m;
    static boolean[] vis = new boolean[N];
    static boolean ok;
    static long[] cnt = new long[3];
    static int[] d = new int[N];
    static int T;
    
    static void dfs(int x, int l) {
        vis[x] = true;
        cnt[l]++;
        d[x] = l;
        for (int i = u[x]; i != 0; i = e[i].pre) {
            T++;
            if (!vis[e[i].to]) {
                dfs(e[i].to, (l + e[i].c) % 3);
            } else if ((l + e[i].c) % 3 != d[e[i].to]) {
                ok = false;
            }
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        for (int i = 1; i <= m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            ins(a, b, 1);
            ins(b, a, 2);
        }
        
        long ans = 0;
        for (int i = 1; i <= n; i++) {
            if (!vis[i]) {
                T = 0;
                for (int j = 0; j <= 2; j++) {
                    cnt[j] = 0;
                }
                ok = true;
                dfs(i, 0);
                long s = cnt[0] + cnt[1] + cnt[2];
                if (!ok) {
                    ans += s * s;
                } else {
                    if (cnt[0] > 0 && cnt[1] > 0 && cnt[2] > 0) {
                        ans += cnt[0] * cnt[1] + cnt[1] * cnt[2] + cnt[2] * cnt[0];
                    } else {
                        ans += T / 2;
                    }
                }
            }
        }
        
        System.out.println(ans);
    }
}